### PR TITLE
Allow Ashe to build ult charge while BOB is active

### LINKED ISSUE #98: Allow Ashe to build ult charge while BOB is active

### PR DESCRIPTION
**Pull request by novalderya**
Fix #98 
Implementation based on https://overwatch.fandom.com/wiki/Ultimate_ability for the charge per second and logic of the calcul of the percentage of ult per damage.

Logic : Add to a variable the number of damage made by Ashe and then add the ult charge when bob dissapear.

Specific case handled: 
    - Ashe tnt herself
    - Switch heroes after using BOB
    
 Example: 
ashe_charge_while_ult
Damage made in the example : 
Capture

Give (708 + 50)/ 2240 = 34  %

50 from the ten seconds BOB is alive.

### CODE REVIEW

File: src/constants/common_constants.opy [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!mainFile "../main.opy"
+
+#!define ULT_CHARGE_PER_SECOND 5
```
Comment (MaxwellJung) [inline src/constants/common_constants.opy:3]: Move this constant to just `ow1_constants.opy`

File: src/main.opy [hunk 2]
```diff
@@ -3,6 +3,7 @@
 # Useful constants
 #!include "constants/ow1_constants.opy"
 #!include "constants/ow2_constants.opy"
+#!include "constants/common_constants.opy"
```
Comment (MaxwellJung) [inline src/main.opy:6]:
```suggestion
```

File: src/ow1/heroes/ashe.opy [hunk 3]
```diff
@@ -14,4 +16,29 @@ rule "[ashe.opy]: Detect Ashe initialization":
...
+ 
+
+
+rule "[ashe.opy]: Calcul of the damage made while BOB is active":
+    @Event playerDealtDamage
+    @Hero ashe
+    @Condition eventPlayer.isUsingUltimate() == true
+    @Condition victim != attacker
+
+    eventPlayer.dmg_done_while_ult_active += eventDamage
```
Comment (MaxwellJung) [inline src/ow1/heroes/ashe.opy:44]: Can you rewrite this using the `missing_ult_points` variable from `src/ow1/passives/no_ult_negation.opy`?

Your idea is solid, but we were planning to use the logic from that file to fix this issue. This works, but it slightly deviates from the ult charge compensation logic defined in that file.

**Review by MaxwellJung — CHANGES_REQUESTED** [review]
Please adapt your logic to fit the ult charge compensation logic defined in `src/ow1/passives/no_ult_negation.opy`.

I probably should have commented to use the variables and rules in that file to fix this issue, but I forgot.

File: src/ow1/heroes/ashe.opy (hunk 3 above)
Comment (novalderya) [reply src/ow1/heroes/ashe.opy]: Implemented with `missing_ult_points`

**Comment by novalderya** [comment]
I got a question regarding this line : https://github.com/MaxwellJung/ow1_emulator/blob/staging/src/ow1/passives/no_ult_negation.opy#L65C126-L65C126
Why do we add 3% which is a lot whenever we want to compensate missing ult point 

For example for ashe during the ultimate 

I do one head shot on a tank so 150*1.3 = 195 ultimate points
10 seconds for BOB : 50 points charge 
So it should give 245/2240 = 10.9 % so 11% 

But with the 3% percent in the end we get 14% percent instead which is not the expected percentage

**Comment by MaxwellJung** [comment]
> I got a question regarding this line : https://github.com/MaxwellJung/ow1_emulator/blob/staging/src/ow1/passives/no_ult_negation.opy#L65C126-L65C126 Why do we add 3% which is a lot whenever we want to compensate missing ult point
> 
> For example for ashe during the ultimate
> 
> I do one head shot on a tank so 150*1.3 = 195 ultimate points 10 seconds for BOB : 50 points charge So it should give 245/2240 = 10.9 % so 11%
> 
> But with the 3% percent in the end we get 14% instead which is not the expected percentage

Honestly, that 3% has no good explanation. It's really just an arbitrary value used to cover up the difference between expected behavior and observed behavior when executing that code in testing.

For example, the code in theory should work as intended without that 3%, but when you actually execute that code by shooting tanks in practice range, you always get less ult charge than shooting non-tanks (like ~10% less ult charge overall for the same amount of damage). I arbitrarily kepting increasing that value, until the expected behavior matched observed behavior.

So, yes, that is technically a bug that someone should fix.

**Review by MaxwellJung — APPROVED** [review]
Looks good; thanks for contributing!

**Pull request merged**